### DISCUSSION TITLE
How to transfer ERC 721 Token to a Reach Smart Contract

### DISCUSSION THREAD

**Discussion by asolpshinning** [reach-sh/reach-lang#1304] [Q&A]
I wrote a contract in Reach to receive ERC 721, but I found out the transaction was being reverted. I researched the reason for it and discovered for any smart contract to receive ERC 721, it must implement this interface: 
[IERC721Receiver.onERC721Received](https://docs.openzeppelin.com/contracts/4.x/api/token/erc721#IERC721Receiver-onERC721Received-address-address-uint256-bytes-), which is called upon a safe transfer.

Wondering if I need to, or how to write this interface in Reach as an API. If I write as an API, it says the function has to return the solidity selector, which I am not sure how to implement that in the Reach function/API...
This is how it looks in solidity:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/IERC721Receiver.sol

Is there an equivalent to the `selector` function in Reach ?

```
function onERC721Received(address, address, uint256, bytes calldata) external pure returns (bytes4) {
        return IERC721Receiver.onERC721Received.selector;
    }
```

Any ideas?

**Comment by jeapostrophe** [comment]
We don't support that API form right now, because we have no types like `bytes` (that uses unbounded space). I will work on it and reply here when I have a solution.

**Reply by asolpshinning** [reply]
Thank you Jay. Pending the time we get a solution, I would love to write the contract with solidity code and deploy it with Reach so I can use Reach backend (compiled JavaScript), to deploy and interact with the contract.
Is this possible ?
If yes, any example code I can look at?
Thanks for your help.

**Reply by jeapostrophe** [reply]
You could write it directly in Solidity, let's call that `Raw`, and have a Reach program interact with it, let's call that `Refined`, by having a `remote` object in `Refined` that is bound to `Raw`. For an example of doing stuff like this, check out --- https://github.com/reach-sh/reach-lang/tree/master/examples/remote

**Comment by willghatch** [comment]
Just FYI, `BytesDyn` is available now: https://docs.reach.sh/rsh/compute/#rsh_BytesDyn

So you can write an API with this signature to support receiving ERC721s: `onERC721Received: Fun([Address, Address, UInt, BytesDyn], Bytes(4))`
